File: 2.add-two-numbers.cpp
/*
 * @lc app=leetcode id=2 lang=cpp
 *
 * [2] Add Two Numbers
 */

#include <bits/stdc++.h>

#include <cassert>

using namespace std;

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next){};
};

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    ListNode* ret = nullptr;
    ListNode* last = nullptr;
    int carry = 0;
    while (1) {
      int sum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
      if (last == nullptr) {
        last = new ListNode(sum % 10);
        ret = last;
        if (!(sum || l1 || l2)) {
          break;
        }
      } else {
        if (!(sum || l1 || l2)) {
          break;
        }
        last->next = new ListNode(sum % 10);
        last = last->next;
      }

      carry = sum / 10;
      if (l1) l1 = l1->next;
      if (l2) l2 = l2->next;
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  assert(0);
  return 0;
}

File: 2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if l1 is None and l2 is None:
            return None

        node1 = l1
        node2 = l2

        ret = None

        carry = 0
        while node1 is not None or node2 is not None or carry != 0:
            sum = 0
            if node1 is not None:
                sum += node1.val
            if node2 is not None:
                sum += node2.val
            sum += carry
        
            carry = 0
            if sum >= 10:
                carry = sum // 10
            tmp = ListNode(sum % 10)
            if ret is None:
                ret = tmp
            else:
                rnode.next = tmp
            rnode = tmp

            if node1 is not None:
                node1 = node1.next
            if node2 is not None:
                node2 = node2.next

        return ret

        
# @lc code=end

